### DISCUSSION TITLE
Equivalent to x509cert-lookup SPI?

### DISCUSSION THREAD

**Discussion by tanadeau** [keycloak/keycloak#10091] [Archive]
We are currently using the x509cert-lookup SPI in the WildFly distribution as we're using NGINX for SSL termination and getting the client certificate. Is there an equivalent in the Quarkus distribution? If so, how is it configured?

Thanks!

**Comment by pedroigor** [comment]
@tanadeau Hi, we are going to change the doc accordingly but the format for SPI related options is : `spi-<spi>-<provider-id>-<property>`.

In this case, it would be:

```
spi-x509cert-lookup-provider=haproxy
```

As you are enabling a specific provider, this configuration is going to take affect only after running the `build` command:

```
kc.sh build --spi-x509cert-lookup-provider=haproxy <more_options>
```

Or if you have set it in `keycloak.conf` or as a env var:

```
kc.sh build
```

**Reply by tanadeau** [reply]
Excellent. Thanks for the info. Would it be similar for custom SPIs? We have an event listener SPI for instance.

**Reply by tanadeau** [reply]
I don't see much documentation about migrating SPI implementations including building, deploying, configuration, etc. with the Quarkus distribution.

**Reply by tanadeau** [reply]
Also some of our SPI properties are currently read from the environment like AWS region, SQS queue URL, etc. Can those still be read outside of the `build` step?

**Reply by pedroigor** [reply]
> I don't see much documentation about migrating SPI implementations including building, deploying, configuration, etc. with the Quarkus distribution.

The only reference we have is https://www.keycloak.org/migration/migrating-to-quarkus. You can check there a little information about the differences when building/deploying.

In regards to configuration, I'm updating docs for now here https://github.com/keycloak/keycloak-documentation/pull/1408.

But I think we should come up with a less verbose configuration when enabling a specific proxy vendor. For instance, we could have:

```
proxy-vendor=haproxy|apache|nginx
```

To automatically configure the server for a specific proxy vendor.

Perhaps we should also update our [Proxy Guide](https://www.keycloak.org/server/reverseproxy) with the information you are requesting. 

@DGuhr What do you think?

**Reply by stianst** [reply]
@pedroigor I would like the documentation of reverse proxy setup around mtls to be moved to the new guides for Quarkus, and not exist in the server admin guides, as it's about configuring the server. +1 with introducing a high-level option around this, not sure about proxy-vendor, it conflicts a bit with db=vendor, and maybe it should be limited to just mtls since that's what it's doing? Haven't looked at the difference and if it really needs to be haproxy/haproxy/ngingx or should instead be something more generic on how to load cert info.

**Reply by DGuhr** [reply]
perhaps sth like `proxy-mtls-conf=...` or `proxy-x509-default-mtls-conf=...` ? Not sure either. A generic solution would be better for sure, haven't looked into the details, yet.

**Comment by pedroigor** [comment]
@tanadeau This discussion should be solved by https://github.com/keycloak/keycloak/pull/10220.

@stianst @DGuhr Created https://github.com/keycloak/keycloak/discussions/10264 to discuss improvements to the proxy configuration category.
